### PR TITLE
AP-243 Integrate RevenueCat

### PR DESCRIPTION
**Pull request by jayeshsolanki93**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -132,6 +144,7 @@ repositories {
     maven {
         url "https://maven.google.com"
     }
+    mavenCentral()
```
Comment (ravjit-cliqz) [inline app/build.gradle]: we already have jcenter in the project level buildscript

File: app/src/lumen/java/com/cliqz/browser/subscription/PurchaseFragment.kt [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    override fun onBuyClicked(position: Int) {
+        mAdapter.getProduct(position)?.apply {
+            Purchases.sharedInstance.makePurchaseWith(activity as Activity, sku,
+                    BillingClient.SkuType.SUBS,
+                    { error, _ ->
+                        Log.e(TAG, error.underlyingErrorMessage)
+                    },
+                    { _, purchaseInfo ->
+                        if (purchaseInfo.allPurchasedSkus.any { it.contains("vpn") }) {
```
Comment (ravjit-cliqz) [inline app/src/lumen/java/com/cliqz/browser/subscription/PurchaseFragment.kt]: match full names and use constants

File: app/src/lumen/java/com/cliqz/browser/subscription/PurchaseFragment.kt [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                if (productList.isEmpty()) {
+                    // TODO: Display error
+                } else {
+                    mAdapter.updateProductList(productList)
+                    setLoading(false)
+                }
+            }
+
+            override fun onError(error: PurchasesError) {
+                Log.w(TAG, error.message)
```
Comment (ravjit-cliqz) [inline app/src/lumen/java/com/cliqz/browser/subscription/PurchaseFragment.kt]: Log.e

File: app/src/lumen/java/com/cliqz/browser/subscription/TrialPeriodHandler.kt [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import java.net.URL
+
+private const val CONTENT_TYPE_JSON = "application/json"
+private val LUMEN_CREDENTIAL_URL = URL("https://auth-staging.lumenbrowser.com/get_credentials")
+
+private val HEADERS = mapOf("x-api-key" to BuildConfig.VPN_API_KEY)
+
+private val REQUEST_BODY = JSONObject()
+        .put("device_id", Purchases.sharedInstance.appUserID)
+        .put("revenue_cat_token", BuildConfig.REVENUECAT_API_KEY)
```
Comment (ravjit-cliqz) [inline app/src/lumen/java/com/cliqz/browser/subscription/TrialPeriodHandler.kt]: This should be the per user revenue cat token returned by revenue cat, not the api-key.

File: app/src/main/java/com/cliqz/browser/main/MainActivity.java [hunk 5]
```diff
@@ -242,6 +257,29 @@ protected void onCreate(Bundle savedInstanceState) {
...
 
+        if ("lumen".equals(BuildConfig.FLAVOR_api)) {
+            // Check if subscribed.
+            Purchases.getSharedInstance().getPurchaserInfo(new ReceivePurchaserInfoListener() {
+                @Override
+                public void onReceived(@androidx.annotation.NonNull PurchaserInfo purchaserInfo) {
+                    if (!purchaserInfo.getActiveSubscriptions().isEmpty()) {
+                        // If subscribed, enable features.
+                        for (String sku : purchaserInfo.getActiveSubscriptions()) {
+                            premiumSubscriptionManager.setVpnEnabled(sku.contains("vpn"));
```
Comment (ravjit-cliqz) [inline app/src/main/java/com/cliqz/browser/main/MainActivity.java]: full string match please

File: app/src/lumen/java/com/cliqz/browser/subscription/PurchaseFragment.kt (hunk 2 above)
Comment (jayeshsolanki93) [reply app/src/lumen/java/com/cliqz/browser/subscription/PurchaseFragment.kt]: Contains made sense here.

These are the products I use for staging:

com.cliqz.android.lumen.staging.sale.basic_vpn
com.cliqz.android.lumen.staging.sale.vpn
com.cliqz.android.lumen.staging.sale.basic

Even these are not the finalized identifiers for the products.

File: app/src/lumen/java/com/cliqz/browser/subscription/TrialPeriodHandler.kt (hunk 4 above)
Comment (jayeshsolanki93) [reply app/src/lumen/java/com/cliqz/browser/subscription/TrialPeriodHandler.kt]: done

File: app/src/lumen/java/com/cliqz/browser/subscription/PurchaseFragment.kt (hunk 3 above)
Comment (jayeshsolanki93) [reply app/src/lumen/java/com/cliqz/browser/subscription/PurchaseFragment.kt]: done

File: app/src/lumen/java/com/cliqz/browser/subscription/PurchaseFragment.kt (hunk 2 above)
Comment (ravjit-cliqz) [reply app/src/lumen/java/com/cliqz/browser/subscription/PurchaseFragment.kt]: > Even these are not the finalized identifiers for the products.
Thats why we should match full strings. So we don't have to change the code if "contains" logic doesnt stand valid anymore

**Pull request merged**